### ISSUE TITLE
Enhancement: allow re-download of subscribed shows without plist editing

### ISSUE THREAD

**Issue by ryanchild** [mackworth/cTiVo#196]
It would be cool if cTiVo tracked previously recorded shows in a subscription by file path in addition to episode ID.  That way, if I came across a recording in Plex that was partial or had bad signal strength, I can simply delete it from within Plex rather than having to go and find the episode in the prevRecorded array in my cTiVo.plist.  Maybe could even have cTiVo remember both the full path (DownloadDirectory + relative/path/to/video/file) and the relative path, so that it still works if the DownloadDirectory was changed since the show was downloaded.

i.e. instead of this:

```
        if ([prevShow[@"episodeID"] isEqualToString: thisShowID]) {
            DDLogVerbose(@"Already recorded: %@ ",prevShow);
            return NO;
        }
```

do something like this (would also need to add something to remove the prevShow item for the show that was deleted)

```
        if ([prevShow[@"episodeID"] isEqualToString: thisShowID] && [[NSFileManager defaultManager] fileExistsAtPath:[prevShow[@"path"]]]) {
            DDLogVerbose(@"Already recorded: %@ ",prevShow);
            return NO;
        }
```

**Comment by mackworth** [comment]
So, I see your issue, but I should point out that's a very different use case. Currently the idea is that you keep it on the Mac as long as you want it. Thus when you've watched it and chosen to delete, it would be rather contrarian to promptly restore it, and you can see that we go to some effort not to do that. (BTW, we do use the file system's metadata so we can get a list of all shows that are currently on the disk; that's how we mark which shows are avail.)

On the other hand, the scenario you're looking at is more "I'm collecting all the XYZ shows, so please download anything I don't have." Maybe this is a different mode for Subscriptions. Then you could just delete bad ones from Plex, and when TiVo records the show again, we will restore the show.  So

An alternative might be to have a contextual menu item in the Subscription table that says something like "Forget Previously Recorded", which simply erases the previousShow list, so that the Subscription reload any future shows (probably not including the currently available ones, as you could manually do those).

That maybe suggests the possibility of a mode that says "Subscriptions don't download any episodes I already have", so it checks to see if the episode's filename already exists before bothering to download.

**Comment by ryanchild** [comment]
Yeah good point, I guess this would be a different mode rather than replacing the existing functionality.

This kind of reminds me of TiVo's "Record: New" and "Record: New and Repeats" OnePass options.  In TiVo's case, it's looking at the air date to determine whether to record the show.  In cTiVo's case, we're looking at whether the show was previously downloaded.  But other than that I think "New" / "New and Repeats" might translate well to cTiVo and would have the benefit of cTiVo's user base already being familiar with the concept from the TiVo UI.

New: Download new recordings if they don't exist on disk and weren't previously downloaded
New and Repeats: Download new recordings if they don't exist on disk

For the most part I think this subscription option would go hand in hand with the OnePass:

* User is recording the new season of a show and wants to delete shows on her mac and/or TiVo as she watches them.  "New" would be selected for both the OnePass on the TiVo and for the subscription on cTiVo
* User is collecting all episodes of a show, "New and Repeats" would be selected for both the OnePass on the TiVo and for the subscription on cTiVo.

On the other hand I'm sure I'm missing several other use cases and maybe re-using the OnePass option name would be more confusing than helpful.

I guess I should clearly state here how I was picturing using cTiVo subscriptions since I kind of half explained it here and half in #195.

I set the TiVo to record all episodes of a show, only keeping at most 5 (or whatever the minimum number is) episodes on the TiVo, to maximize DVR space.  Whenever cTiVo sees an episode that I don't have on the mac, it downloads it before the next episode is recorded on the TiVo and old recordings are deleted.

The key phrase here is "don't have on the mac" ...I mean this in the literal sense in that it doesn't exist on disk.  I think I would also need an option like "Remove successful downloads from the queue" if it doesn't already exist since if I understand correctly cTiVo won't re-download something (even episodes of subscriptions?) that's still in the download queue.

Anyway, if you're willing to pursue this, I'm open to any implementation you choose.  Or if this is low on your list maybe I can even take a shot at implementing it.  Seems a bit daunting though just because of the way the filesystem / download queue / prevRecorded methods of tracking seem to be somewhat at odds with each other.

**Comment by mackworth** [comment]
Well, Subscriptions are all based on the prevRecorded mechanism.

The restriction on the download queue not having duplicate shows is one I've just never gotten around to. There's no reason it shouldn't work now, just testing/debugging as there may be places we've assumed that restriction is in place. (As an example, parts of Subscriptions itself may assume that it'll be protected from duplicating existing show. Or not.)

The other reason is that when I start thinking about it, it takes me down a more difficult path as one tries to optimize it. Having allowed that, do you then optimize the pipeline? Do you look ahead and say "Oh, I've got three downloads (in different Formats) for this show, so I'll download/decrypt/skip/caption it once and then feed it through the others"? Then once you can run a Format from the file then folks would like to be able to run cTiVo on existing files without a corresponding Show on the TiVo...  So the psychic burden has slowed me from starting that path.

The filesystem "showsOnDisk" is simply to (a) boldface those shows that are already on disk and (b) to do the Show File/Play Video commands and (c) to supply a filename for drag/drop . Not used today at all in Subscriptions or downloading.

Well, I like the idea, but regardless, I want to get 2.5 out. With all the big and little things added and fixed (many found with your prompting and help), the current one is far better than the "shipping" 2.4.4, so any major new features should probably wait for next pass.

**Comment by ryanchild** [comment]
Rather than allowing the download queue to have duplicate shows maybe just add an option to automatically remove from queue when complete?  That would eliminate the scenario you mentioned and would probably require the least amount of changes to the rest of cTiVo.

Agreed, this will take some thought, let's shelve for now, I don't want to distract from the 2.5 release.

**Comment by ryanchild** [comment]
Just a heads up, my debug builds are crashing after beginning a download after upgrading to Xcode 8 with EXC_BAD_ACCESS, not sure yet if something changed with Cocoa or if it's user error.

**Comment by mackworth** [comment]
Have you tried the official beta (built w/ 8, as has all of mine for the last couple months, beta then final a few days ago?

**Comment by ryanchild** [comment]
Your beta release runs fine, but if I pull 2.5 and build the Debug target I get the crash after upgrading Xcode.  No local changes, other than changing the signing.  Since you're not seeing the same thing I guess I must be missing some setting or dependency, I'll keep poking at it...

**Comment by mackworth** [comment]
instantly upon running? can't even set a breakpoint in didLaunchApplication?

**Comment by ryanchild** [comment]
Nope it was happening after processing subscriptions and beginning the download of a new episode.  Kicked me to main.c without any kind of easily recognizable (at least by me) stack trace.

**Comment by mackworth** [comment]
Did this ever get resolved?

**Comment by ryanchild** [comment]
The debug crash?  No, I haven't had time to look into it much.  But since the 2.5 beta release works fine I haven't seen a need to escalate this, basically I'm just blocked from doing any "dev work" if you can call it that (really just poking at the source), beyond enabling zombie objects which is pretty slow.  But again I haven't had a lot of time to play with the cTiVo source lately anyway (though I remain an active user).  

I doubt you want to support my debug builds but in case you want to track this crash I've filed #202, let me know over there if I can provide any additional info.  Or feel free to close until this shows up for someone in an actual release, or you start seeing it in your debug builds.

**Comment by mackworth** [comment]
So ... reading back through all that, I have no idea where we ended up. There is a "hidden" preference for duplicates `write com.cTiVo.ctivo AllowDups YES`. So, assuming you're still out there, was there something else you were proposing here or ... ?

**Comment by ryanchild** [comment]
Thanks, trying to test but my mac mini is having issues.  So what happens if AllowDups is YES and SubscriptionExpirationDays is 100 and cTiVo finds a newly recorded episode in a subscription that was previously downloaded within the last 100 days but no longer exists on disk?  It automatically downloads it?  Or AllowDups just lets you download a non-subscription show that's already in the download list?  I've always just removed the show from the download list before downloading, to do that.

**Comment by mackworth** [comment]
AllowDups is more for people who want to use alternative Formats with a download. You might want a small version to copy onto phone, and a big one for archive/Plex/AppleTV.

So, no it has no effect on re-downloading subscriptions that have already downloaded, but are now missing from disk. Is that the goal? (I've honestly forgotten/can't pick up the thread from above.)

**Comment by ryanchild** [comment]
I see.  Yeah reading back through the thread I think I was imagining being able to just download whatever is on the TiVo, if it doesn't already exist on disk.  Then I could set and forget all shows to only keep 3 episodes and use a small hard disk on the TiVo, compressing and transferring everything to a disk array on the network for unlimited storage + redundancy compared to a TiVo Roamio OTA maxed out at 2TB.

That would almost work with cTiVo today except that whether shows are downloaded depends on whether the show is in the download shows list, and I think the subscriptions feature remembers what it's downloaded via plist.  If a show didn't actually get recorded as intended (bad antenna signal or bad guide data) you'd have to go figure out how to enable cTiVo to re-download it rather than it happening automatically after the file is deleted.  Plus the downloaded shows list would get really long..

Anyway I'm rambling, I'm not even sure what this feature would look like in cTiVo anymore or if it's possible to implement cleanly.  I think when I filed this I was just hoping cTiVo could be made to download subscription shows based on whether the show is on disk rather than on the list of previously downloaded shows.  But I forgot about the downloaded shows list..

Really feel free to just close this if I'm not making sense or you feel this sort of thing is outside the scope of cTiVo, it's already a very versatile and useful tool, keep up the good work as always!

**Comment by mackworth** [comment]
So, sounds like you're looking for an option to just use existence on hard disk instead of the download list. While that wouldn't work for many folks, who delete shows after viewing, so they don't want it to download again, I could easily see a hidden option simply ignores the previously-recorded list and uses existence on hard disk instead. Secondly, if you have a bad recording, then you could just delete it from your disk and if TiVo records it again, we would automatically pick it up. (If you delete from your disk and it's still on TiVo, it would download again next time you quit/restart cTiVo OR if you right-click Reapply To Shows in the Subscription menu. )

That would be easy, but it means that there's no way for you to delete a show off your hard disk while it's still on the TiVo.

**Comment by mackworth** [comment]
So, I stuck this into 3.1.1 at the last second, and I've only tested that it doesn't interfere with regular usage. If you exit cTiVo, wait a few seconds, and then use Terminal to  `defaults write com.ctivo.ctivo SubscriptionRelyOnDiskOnly YES`, it should rely solely on whether it can see a subscription.  `defaults delete com.ctivo.ctivo SubscriptionRelyOnDiskOnly` to go back to normal.

So, give it a try under controlled circumstances; if it does what you want, great!

**Comment by mackworth** [comment]
Have you had a chance to try this out?

**Comment by ryanchild** [comment]
Actually I've had this set since you released 3.1.1 and it's still re-downloading shows that exist on disk but weren't downloaded in the last 90 days.  I realize I can set SubscriptionExpirationDays 1000 or whatever to get fewer duplicates but I was thinking this SubscriptionRelyOnDiskOnly YES setting would skip the whole prevRecorded mechanism and just check the disk?  I also tried deleting a subscription for which a show download recently occurred, renamed the downloaded file to something else, re-subscribed, and it wasn't queued up for download like I would expect.  However if I delete the show from the download queue and re-apply the subscription to shows, it is queued again for download whether the file exists on disk or not.

So I guess I could use some clarification on exactly what the expected behavior of this new hidden SubscriptionRelyOnDiskOnly setting is?

Again I'm just trying to get it to download whatever's on the tivo to disk if it's not already there, regardless of whether or when it was previously downloaded, or whether it's in the download queue.

**Comment by mackworth** [comment]
So, the way this is supposed to work (code snippet below) is that that preference simply ignores the previously recorded list entirely, and looks whether a potential show is marked on disk. So if you see a checkmark in the OnDisk column in the NowPlaying table (the top one), then it won't record it, otherwise it should. This should be true for all subscriptions, whether you create a new one or not.

Now that I write that, I realize maybe you're talking about when TiVo itself does a second recording of the same episode of the show? That's a problem, as we mark the shows on disk with "TivoName-TivoID", not with the episodeID, whereas the normal subscription list tracks the episodeID. (TiVoID is unique to a particular recording on a TiVo, while EpisodeID is the same for both recordings.) I realize that's probably not what you're looking for, and will have to think more about whether there's still a way to make it work.

>  it's still re-downloading shows that exist on disk but weren't downloaded in the last 90 days.

So, if it's a second recording, that's true, and in fact it will do so even if the first one was downloaded just recently.

> also tried deleting a subscription for which a show download recently occurred, renamed the downloaded file to something else, re-subscribed, and it wasn't queued up.

Renaming or even moving the file doesn't do anything. You'd have to remove the metadata in Terminal for this test to work.

>  if I delete the show from the download queue and re-apply the subscription to shows, it is queued again for download whether the file exists on disk or not.

This one is surprising.  Can you confirm that there's a checkmark in the OnDisk column for that show?

Note that if you turn on Verbose logging mode for Subscription in Adv Prefs, you'll get more detailed information on what's happening.
Similarly, if you turn on Verbose for TivoManager, you'll get the list of shows that it "sees" on disk (look for "showsOnDisk" in log).

````ObjectiveC
if ([[NSUserDefaults standardUserDefaults] boolForKey:kMTSubscriptionRelyOnDiskOnly]) {
	if (tivoShow.isOnDisk) {
		DDLogVerbose(@"Found on disk: %@ ",tivoShow);
		return NO;
	}
} else {
	NSString * thisShowID = tivoShow.uniqueID; //uniqueID = episodeID for episodic shows; adds original air date for non-episode ones.
	for (NSDictionary * prevShow in self.prevRecorded ) {
		if ([prevShow[@"episodeID"] isEqualToString: thisShowID]) {
			DDLogVerbose(@"Already recorded: %@ ",prevShow);
			return NO;
		}
	}
}
return YES;
````

**Comment by ryanchild** [comment]
> This one is surprising. Can you confirm that there's a checkmark in the OnDisk column for that show?

There is not.  In fact nothing in the NowPlaying table has OnDisk checked, even shows that were downloaded and can be found on disk by right clicking on the download queue entry and selecting "show in finder"
I'm not sure why this particular show wouldn't be OnDisk, since it was first aired a couple weeks ago so I don't think the show that was downloaded had a different TivoID than the one in NowPlaying (should be the same thing on the tivo)

Before we go much further down this rabbit hole I should note that I'm working on moving my home server to linux or windows and so will probably be using kmttg or some custom script for my tivo downloading needs going forward.  I really appreciate all the work you've put into this though and want to help with testing as best I can.  Just something to consider before pouring more energy into this issue.. I am OK with closing at any time.

**Comment by mackworth** [comment]
So, I can’t remember if we looked into this before. That probably says your disk isn’t saving metadata. What format is it?  (NFS, in particular, doesn’t support it). 

Note that “Show in Finder” for Downloads use a different mechanism (they remember their path, so if you have two recordings of the same show, it will take you to the right one, whereas Show in Finder for Shows takes you to all previous recordings.)

`xattr -l path` will show you if a file has any attributes.
`xattr -w attrName attrValue path` lets you set one.

**Comment by ryanchild** [comment]
Hmm the metadata appears to be getting saved.  It's HFS+ journaled.
It looks like the reason most of my downloads were missing OnDisk is that it gets lost when I quit and restart cTiVo, is that a bug?

Anyway you're correct, I was thinking of a situation where the tivo re-records a show which is going to be quite common for onepasses that include repeats.  It sounds like setting SubscriptionExpirationDays 1000 in that case won't result in duplicates (for a few years) but SubscriptionRelyOnDiskOnly YES will, since it sees different TivoIDs for each recording.

Still wondering what the best way to test your new setting is... if I delete a download for a show in finder that's still on the tivo, and then re-apply the subscription to shows, it should start a new download?  Deleting the download from the download queue still seems to be necessary in addition to deleting in finder, maybe that's because I don't have AllowDups set?

**Comment by mackworth** [comment]
Okay, that's definitely weird.  If you do a search in Finder for `cTiVoDownload` when cTiVo isn't running, you see the shows, but when you start cTiVo, they're not marked?

Can you set TiVoManager log level to Verbose, and then quit/restart ctVo? The Log should have an entry of `showsOnDisk = ` followed by a list of all the shows on disk.

So the new setting should do the opposite, if you reapply the subscription, while the show is still on disk, it should NOT create the download.  But if you then delete from finder and do it again, it should recreate it.  Yes, if you don't have AllowDups, then you'll also need to delete the original download, or it won't create a second one.

**Comment by ryanchild** [comment]
it only shows one show on disk:
```
2018-05-07 07:08:55:360 MTTiVoManager>metadataQueryHandler:@190>showsOnDisk = {
    "Living Room: 9723" =     (
        "/Users/ryan/Desktop/comskip/Fresh Off the Boat - S01E04 - Success Perm.ts"
    );
}
```
That is an old download folder, right now it's set to /Volumes/Storage/DVR and that's where all the shows are, looks like cTiVo isn't finding them for some reason.

Yeah sorry, typo, I guess I meant "if I delete a download for a show in finder that's still on the tivo, and then re-apply the subscription to shows, it SHOULDN'T start a new download"
But only if the recording on the tivo was downloaded before, and it's not a new recording of a show it's downloaded before, because the TivoID will be different ...right?

**Comment by ryanchild** [comment]
```
bernardel:~ ryan$ xattr -l /Users/ryan/Desktop/comskip/Fresh\ Off\ the\ Boat\ -\ S01E04\ -\ Success\ Perm.ts 
TiVoID: 9723
TiVoName: Living Room
com.apple.metadata:kMDItemFinderComment: cTiVoDownload
com.apple.quarantine: 0002;57aa223a;QuickTime Player;
bernardel:~ ryan$ xattr -l /Volumes/Storage/DVR/TV\ Shows/Modern\ Family/Season\ 09/Modern\ Family\ -\ S09E20\ -\ Mother\!.mp4 
TiVoID: 128176
TiVoName: Living Room
com.apple.metadata:kMDItemFinderComment: cTiVoDownload
```

Finder finds the first file but not the second.  It looks like the attributes are already correctly set on both.  Are you using some kind of file discovery thing to find the files?  I have /Volumes/Storage disabled under settings -> spotlight -> privacy, this drive has tons of files and I think I was trying to disable indexing for performance since I really just use it for archive and plex.  But I can't remember whether it really helped, maybe I can re-enable /Volumes/Storage for spotlight if cTiVo requires it.  Maybe you want to add a warning or something if the download directory is disabled in spotlight?

**Comment by mackworth** [comment]
Yup, that’s exactly the issue then. We rely on the OS reporting the metadata, which is reported by the indexer. 

I’ll look into that warning.

**Comment by ryanchild** [comment]
OK, cTiVo shows shows and downloads as OnDisk now, but if I delete one of the downloaded shows of a subscription in Finder, the OnDisk check in the download window goes away, but if I click "re-apply to shows" nothing happens..?  I would expect it to queue up a new download?

**Comment by mackworth** [comment]
Yup. Increase the log level in Subscription to Verbose, and see what it says.

**Comment by ryanchild** [comment]
```
2018-05-09 10:59:42:849 MTTiVoManager>addToDownloadQueue:beforeDownload:@1247> Star Trek: Spock's Brain (Living Room) already in queue
```

but

```
bernardel:~ ryan$ defaults read com.cTiVo.ctivo | grep -e AllowDups -e SubscriptionRelyOnDiskOnly
    AllowDups = YES;
    SubscriptionRelyOnDiskOnly = YES;
```

shouldn't AllowDups override "already in queue?"

**Comment by mackworth** [comment]
Absolutely. I recently can't see how you could get that msg (code below, which is the only place "already in queue" appears.)
I assume you haven't changed AllowDups during ctivo operation.
Shouldn't change anything, but does drag&drop give same error?  Add to Queue button?
Does subscription thenwork if you delete the old download?

````ObjectiveC
if ( ! [[NSUserDefaults standardUserDefaults] boolForKey:kMTAllowDups]) {
	for (MTDownload *oldDownload in _downloadQueue) {
		if (oldDownload.show.showID == newShow.showID	) {
			if (oldDownload.show==newShow) {
				DDLogDetail(@" %@ already in queue",oldDownload);
			} else {
				DDLogReport(@"show %@ already in queue as %@ with same ID %d!",oldDownload, newShow, newShow.showID);
			}
			showFound = YES;
			break;
		}
	}
}
````

**Comment by ryanchild** [comment]
Well I thought I had set it before starting cTiVo but I restarted cTiVo to be sure and it seems to be working now, thanks.  Do you want to close this, or does what I proposed sound interesting, i.e. decide whether to download based on the file name instead of the TivoID, to prevent duplicates from showing up when old episodes are re-aired.

**Comment by ryanchild** [comment]
Actually, when I restarted cTiVo with AllowDups, the same show on the tivo is queued for download 3 times.  3 downloads already exist on disk although OnDisk is not checked in the NowPlaying list.  OnDisk is checked for the 3 earlier completed items in the download queue.  Never quite caught the meaning of OnDisk in the NowPlaying list.  Is it supposed to be checked if any download of that show exists on disk?  I don't see anything obvious in the log but let me know if there's something I should look for.

**Comment by ryanchild** [comment]
Hmm ok, I’m still not sure why OnDisk shows in the download table but not the now playing table, but maybe the issue you mention is one of the reasons for the multiple instances in the download queue.  Let me know if I can do anything else to test with the current or any future releases.

**Comment by mackworth** [comment]
So, I took out the SubscriptionsRelyOnDisk for now as it was causing great grief with dups allowed, but I did find and fix several problems with OnDisk in 3.2.0. Give it a try and let me know what else still you think I should do here.
